failure-injection: add node kill failure

DarrylWong opened the pull request:
This change adds a node kill failure mode which can either gracefully or ungracefully shutdown a cockroach process.

Informs: https://github.com/cockroachdb/cockroach/issues/138970
Release note: none

DarrylWong committed:
failure-injection: add cockroach process kill failure

This change adds a cockroach node kill failure mode which
can either gracefully or ungracefully shutdown a
cockroach process.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/144411)
<!-- Reviewable:end -->

DarrylWong committed:
roachtest/failure-injection: add regex filter for smoke test

This will make it easier to add new failure modes without
running the entire suite.

DarrylWong commented on pkg/cmd/roachtest/tests/failure_injection.go:
```diff
@@ -599,6 +613,97 @@ var dmsetupDiskStallTest = func(c cluster.Cluster) failureSmokeTest {
+			killedNodeGroup = groups[0]
+			unaffectedNodeGroup = groups[1]
+
+			// If there is only one live node left, draining will hang indefinitely while
```
A followup PR will bump the size of this cluster so we can test rebalancing ranges. So this is only temporary that we don't test multi node graceful drains.

srosenberg commented on pkg/roachprod/failureinjection/failures/node_kill.go:
```diff
@@ -0,0 +1,97 @@
+	}, nil
+}
+
+type NodeKillFailure struct {
```
Should this be renamed to ProcessKillFailure to differentiate from an upcoming VMKillFailure? (The latter also has two modes of "shutdown", at least in GCE.)

srosenberg commented on pkg/roachprod/failureinjection/failures/node_kill.go:
```diff
@@ -0,0 +1,97 @@
+	nodes := args.(NodeKillArgs).Nodes
+	signal := unix.SIGKILL
+	if args.(NodeKillArgs).GracefulShutdown {
+		signal = unix.SIGTERM
```
Another complication is we actually use `SIGUSR1` to get coverage instrumentation written out, when enabled.

srosenberg replied:
Why not use `roachprod.Stop` which already supports all of the above?

DarrylWong replied:
I was copying DRT's implementation which switched from a `roachprod.Stop` approach to the pkill approach: https://github.com/cockroachdb/cockroach/pull/123517

It's only an issue for DRT since it skips service registry, so maybe we opt for two modes here? I'll experiment with `roachprod.Stop` though since it's better for tenants as Herko pointed out.

>Historically, this has a high chance of failing to kill the process.

Do you have any context of why it fails to kill the process? I did run into one case where if you kill enough nodes, the drain hangs as theres less/no nodes to transfer range leases to. I suppose I viewed this as an acceptable edge case, since it's not the failure mode misbehaving.

srosenberg replied:
> Do you have any context of why it fails to kill the process? 

Just the vague recollection that the drain path doesn't guarantee process termination; i.e., possibly edge-cases around unbounded waiting.

srosenberg replied:
> Another roadblock with using roachprod.Stop though is that it forces a synchronous wait which goes against how the failure injection framework is set up, i.e. wait is optional based on if WaitForFailureToPropagate is called. 

As Herko suggested, if the issue is merely bridging sync calls with async API, then a helper goroutine (and channel) should suffice, right? What if a hypothetical failure injection command needs to run `rm -rf` on bazel's cache directory; surely that would take a long time :)

herkolategan commented on pkg/roachprod/failureinjection/failures/failure.go:
```diff
@@ -209,6 +209,28 @@ func (f *GenericFailure) WaitForSQLUnavailable(
+) error {
+	start := timeutil.Now()
+	err := retryForDuration(ctx, timeout, func() error {
+		res, err := f.RunWithDetails(ctx, l, node, "systemctl is-active cockroach-system.service")
```
Are we going to consider SQL processes (tenants / virtual clusters) as well?
I think as Stan mentioned we could use `.Stop` which also supports SQL processes.

DarrylWong replied:
"Eventually". For now everything is system only, as I figured it'd be easier to focus on that. Definitely will look into `roachprod.Stop` though.

herkolategan commented:
I'm curious if we shouldn't integrate this with `Monitor`. So that we can notify monitor of an expected failure, and then also use the monitor event channel to wait for the node death.

Although it likely will need some refactoring, for instance we'd need a monitor interface in `roachprod` similar to:
https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/test/test_monitor.go

We'd likely need to change the single channel monitor provides to a broadcast like pattern that can be subscribed to.

I'm just thought dumping, trying to figure how this will end up working with roachtests that are running a monitor and have a failure plan. But alternatively it could probably be decoupled if the FI framework notifies the test framework of a planned node kill, which would then notify the monitor.

DarrylWong committed:
failureinjection: catch runAsync panics as errors

Now that the failureinjection framework can spin up goroutines,
make sure to catch all panics as errors. This keeps it compatible
with the roachtest framework, where we don't want any uncaught
panics to take down an entire run.

DarrylWong commented:
>I'm curious if we shouldn't integrate this with Monitor.

I haven't thought about this part much yet (definitely something I should soon), but my thinking is that it should be integrated one level higher at the FI controller that orchestrates events rather than the lower level FI library. Most if not all failure modes should send an event to the Monitor, so it seems easier for the controller to handle that? 

------

Think this PR is RFAL. I didn't directly use `roachprod.Stop` like suggested, but I extracted out the commands it runs into a common helper and run that as a systemd job. This way we can reuse the same logic but have the wait be async so it adheres to the Failure Mode contract.

srosenberg commented on pkg/roachprod/failureinjection/failures/process_kill.go:
```diff
@@ -0,0 +1,173 @@
+		label,
+		waitCmd,
+	)
+	if err := f.c.PutString(ctx, l, node, killCmd, "/tmp/kill.sh", 0755); err != nil {
```
Maybe use `name` as a suffix for `kill.sh` to eliminate collisions/overwrites?

srosenberg replied:
Technically, we should also clean it up after the systemd unit is done.

srosenberg commented:
> Think this PR is RFAL. I didn't directly use `roachprod.Stop` like suggested, but I extracted out the commands it runs into a common helper and run that as a systemd job. This way we can reuse the same logic but have the wait be async so it adheres to the Failure Mode contract.

It seems ok but now we have two different implementations of killing a node, and both have lurking edge-cases. Is it more involved than meets the eye, to have an async wrapper around `c.kill`?

DarrylWong commented:
Thanks for the comments guys, I was way overthinking it. I originally passed on the goroutine channel approach because it would silently swallow errors unless you called `WaitForFailureToPropagate`. But as Herko pointed out, the systemd approach I went with did the same thing 😄 I don't expect it to, but I can always revisit if it causes any issues.

It's now using `Stop` directly, which has the added benefit of having being compatible with virtual clusters. The actual failure mode implementation is quite bare bones now (in a good way 😃)
